From d0f1d05981f97527061bb9997e9527e1a75d37ee Mon Sep 17 00:00:00 2001
From: David Clark <dclark@sierrawireless.com>
Date: Wed, 7 Mar 2018 12:48:38 -0800
Subject: [PATCH] Fix LwM2MCore processing duplicate message IDs - p2

Updated LwM2MCore to keep track of the last received message ID (and if it
is valid) and only process those message IDs which have not already been
received.

Resolves: LE-9669
Change-Id: I829f87f5511d26cb9a5c8ed5d9b66bc62a5cc7f4
Signed-off-by: David Clark <dclark@sierrawireless.com>
---
 core/liblwm2m.c |  2 ++
 core/liblwm2m.h |  2 ++
 core/packet.c   | 28 ++++++++++++++++++++++++----
 3 files changed, 28 insertions(+), 4 deletions(-)

diff --git a/core/liblwm2m.c b/core/liblwm2m.c
index 667555b..44d9abd 100644
--- a/core/liblwm2m.c
+++ b/core/liblwm2m.c
@@ -74,6 +74,8 @@ lwm2m_context_t * lwm2m_init(void * userData)
         contextP->userData = userData;
         srand((int)lwm2m_gettime());
         contextP->nextMID = rand();
+        contextP->lastRxMID = (uint16_t) - 1;
+        contextP->lastRxMIDValid = false;
     }
 
     return contextP;
diff --git a/core/liblwm2m.h b/core/liblwm2m.h
index 5e274ff..32b032b 100644
--- a/core/liblwm2m.h
+++ b/core/liblwm2m.h
@@ -720,6 +720,8 @@ typedef struct
     void *                     bootstrapUserData;
 #endif
     uint16_t                nextMID;
+    uint16_t                lastRxMID;
+    bool                    lastRxMIDValid;
     lwm2m_transaction_t *   transactionList;
     void *                  userData;
 } lwm2m_context_t;
diff --git a/core/packet.c b/core/packet.c
index 8af7a37..24a1747 100644
--- a/core/packet.c
+++ b/core/packet.c
@@ -148,7 +148,8 @@ static bool is_block_transfer(coap_packet_t * message, uint32_t * block_num, uin
 static uint8_t handle_request(lwm2m_context_t * contextP,
                               void * fromSessionH,
                               coap_packet_t * message,
-                              coap_packet_t * response)
+                              coap_packet_t * response,
+                              uint8_t validMid)
 {
     lwm2m_uri_t * uriP = NULL;
     uint8_t result = COAP_IGNORE;
@@ -196,8 +197,12 @@ static uint8_t handle_request(lwm2m_context_t * contextP,
                     coap_init_message(response, COAP_TYPE_ACK, 0, message->mid);
                     message_send(contextP, response, fromSessionH);
 
-                    // Get actual response from user app
-                    return lwm2mcore_CallCoapEventHandler(message);
+                    LOG_ARG("Valid(%u) mid(%u)", validMid, message->mid);
+                    if (validMid)
+                    {
+                        // Get actual response from user app
+                        return lwm2mcore_CallCoapEventHandler(message);
+                    }
                 }
             }
         }
@@ -420,6 +425,7 @@ void lwm2m_handle_packet(lwm2m_context_t * contextP,
                          int length,
                          void * fromSessionH)
 {
+    uint8_t validMid = false;
     uint8_t coap_error_code = NO_ERROR;
     static coap_packet_t message[1];
     static coap_packet_t response[1];
@@ -433,6 +439,7 @@ void lwm2m_handle_packet(lwm2m_context_t * contextP,
     coap_error_code = coap_parse_message(message, buffer, (uint16_t)length);
     if (coap_error_code == NO_ERROR)
     {
+        LOG_ARG("Last received mid/valid %u/%u", contextP->lastRxMID, contextP->lastRxMIDValid);
         LOG_ARG("Parsed: blk: %u ver %u, type %u, tkl %u, code %u.%.2u, mid %u, Content type: %d",
                 message->block1_num, message->version, message->type, message->token_len, message->code >> 5,
                 message->code & 0x1F, message->mid, message->content_type);
@@ -502,6 +509,13 @@ void lwm2m_handle_packet(lwm2m_context_t * contextP,
                     // handle block 1
                     coap_error_code = coap_block1_handler(&serverP->block1Data, message->mid, message->payload, message->payload_len, block1_size, block1_num, block1_more, &complete_buffer, &complete_buffer_size);
 
+                    if (!block1_more)
+                    {
+                        validMid = (!contextP->lastRxMIDValid || (contextP->lastRxMID + 1 == message->mid));
+                        contextP->lastRxMID = (contextP->lastRxMID != message->mid) ? message->mid:contextP->lastRxMID;
+                        contextP->lastRxMIDValid = (!contextP->lastRxMIDValid) ? true:contextP->lastRxMIDValid;
+                    }
+
                     // if payload is complete, replace it in the coap message.
                     if (coap_error_code == NO_ERROR)
                     {
@@ -518,9 +532,15 @@ void lwm2m_handle_packet(lwm2m_context_t * contextP,
                 coap_error_code = COAP_501_NOT_IMPLEMENTED;
 #endif
             }
+            else
+            {
+                validMid = (!contextP->lastRxMIDValid || (contextP->lastRxMID + 1 == message->mid));
+                contextP->lastRxMID = (contextP->lastRxMID != message->mid) ? message->mid:contextP->lastRxMID;
+                contextP->lastRxMIDValid = (contextP->lastRxMIDValid == false) ? true:contextP->lastRxMIDValid;
+            }
             if (coap_error_code == NO_ERROR)
             {
-                coap_error_code = handle_request(contextP, fromSessionH, message, response);
+                coap_error_code = handle_request(contextP, fromSessionH, message, response, validMid);
             }
             if (coap_error_code==NO_ERROR)
             {
-- 
2.11.0

